pkg/platform/local/platform.go

```diff
@@ -540,92 +524,48 @@ func (p *Platform) deployFunction(createFunctionOptions *platform.CreateFunction
-		"port", functionHTTPPort,
-		"previousHTTPPort", previousHTTPPort)
-
-	labels := map[string]string{
-		"nuclio.io/platform":      "local",
-		"nuclio.io/namespace":     createFunctionOptions.FunctionConfig.Meta.Namespace,
-		"nuclio.io/function-name": createFunctionOptions.FunctionConfig.Meta.Name,
-		"nuclio.io/function-spec": p.encodeFunctionSpec(&createFunctionOptions.FunctionConfig.Spec),
-	}
-
-	for labelName, labelValue := range createFunctionOptions.FunctionConfig.Meta.Labels {
-		labels[labelName] = labelValue
-	}
-
-	marshalledAnnotations := p.marshallAnnotations(createFunctionOptions.FunctionConfig.Meta.Annotations)
-	if marshalledAnnotations != nil {
-		labels["nuclio.io/annotations"] = string(marshalledAnnotations)
-	}
-
-	// create processor configuration at a temporary location unless user specified a configuration
-	localProcessorConfigPath, err := p.createProcessorConfig(createFunctionOptions)
-	if err != nil {
-		return nil, errors.Wrap(err, "Failed to create processor configuration")
-	}
-
-	// create volumes string[string] map for volumes
-	volumesMap := map[string]string{
-		localProcessorConfigPath: path.Join("/", "etc", "nuclio", "config", "processor", "processor.yaml"),
-	}
-
-	for _, volume := range createFunctionOptions.FunctionConfig.Spec.Volumes {
-
-		// only add hostpath volumes
-		if volume.Volume.HostPath != nil {
-			volumesMap[volume.Volume.HostPath.Path] = volume.VolumeMount.MountPath
-		}
-	}
-
-	envMap := map[string]string{}
-	for _, env := range createFunctionOptions.FunctionConfig.Spec.Env {
-		envMap[env.Name] = env.Value
+		return nil, errors.Wrap(err, "Failed to compile volumes map")
 	}
+	labels := p.compileDeployFunctionLabels(createFunctionOptions)
+	envMap := p.compileDeployFunctionEnvMap(createFunctionOptions)
 
 	// run the docker image
-	containerID, err := p.dockerClient.RunContainer(createFunctionOptions.FunctionConfig.Spec.Image, &dockerclient.RunOptions{
+	runContainerOptions := &dockerclient.RunOptions{
 		ContainerName: p.GetContainerNameByCreateFunctionOptions(createFunctionOptions),
-		Ports:         map[int]int{functionHTTPPort: 8080},
+		Ports:         map[int]int{dockerclient.RunOptionsNoPort: abstract.FunctionContainerHTTPPort},
 		Env:           envMap,
 		Labels:        labels,
 		Volumes:       volumesMap,
 		Network:       functionPlatformConfiguration.Network,
 		RestartPolicy: functionPlatformConfiguration.RestartPolicy,
-	})
+	}
 
+	// get function port - either from configuration, from the previous deployment or from a free port
+	functionHTTPPort, err := p.getFunctionHTTPPort(createFunctionOptions, previousHTTPPort)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get function HTTP port")
+	}
+	if functionHTTPPort != 0 {
+		p.Logger.DebugWith("Running container with a specific port",
+			"functionHTTPPort", functionHTTPPort)
+		runContainerOptions.Ports = map[int]int{functionHTTPPort: abstract.FunctionContainerHTTPPort}
+	}
```

Comment:
resolve the functionHTTPPort before compiling runContainerOptions so you only fill up once the correct thing instead of filling and re-writing.
Also, call it what it is - externalHTTPPort or more accurately hostHTTPPort